core/src/test/java/org/apache/calcite/test/TopDownOptTest.java

```diff
@@ -288,6 +290,44 @@
         .removeRule(EnumerableRules.ENUMERABLE_JOIN_RULE)
         .check();
   }
+
+  // Not push down sort for hash join in inner join case.
+  @Test void testHashJoinInnerJoinNotPushDownSort() {
+    final String sql = "select * from\n"
+        + "sales.emp r join sales.bonus s on r.ename=s.ename and r.job=s.job\n"
+        + "order by r.job desc nulls last, r.ename nulls first";
+    Query.create(sql)
+        .removeRule(EnumerableRules.ENUMERABLE_MERGE_JOIN_RULE)
+        .removeRule(EnumerableRules.ENUMERABLE_BATCH_NESTED_LOOP_JOIN_RULE)
+        .check();
+  }
+
+  // Not push down sort for hash join in full outer join case.
+  @Test void testHashJoinFullOuterJoinNotPushDownSort() {
+    final String sql = "select * from\n"
+        + "sales.emp r full outer join sales.bonus s on r.ename=s.ename and r.job=s.job\n"
+        + "order by r.job desc nulls last, r.ename nulls first";
+    Query.create(sql)
+        .removeRule(EnumerableRules.ENUMERABLE_MERGE_JOIN_RULE)
+        .removeRule(EnumerableRules.ENUMERABLE_BATCH_NESTED_LOOP_JOIN_RULE)
+        .check();
+  }
+
+  // Push down sort to left input.
+  @Test void testHashJoinLeftOuterJoinPushDownSort() {
+    final String sql = "select * from\n"
+        + "(select ename from sales.emp UNION ALL select ename from sales.emp UNION ALL select ename from sales.emp UNION ALL select ename from sales.emp UNION ALL select ename from sales.emp UNION ALL select ename from sales.emp UNION ALL select ename from sales.emp UNION ALL select ename from sales.emp UNION ALL select ename from sales.emp) r left outer join\n"
+        + "(select ename from sales.bonus UNION ALL select ename from sales.bonus UNION ALL select ename from sales.bonus UNION ALL select ename from sales.bonus UNION ALL select ename from sales.bonus UNION ALL select ename from sales.bonus UNION ALL select ename from sales.bonus UNION ALL select ename from sales.bonus UNION ALL select ename from sales.bonus UNION ALL select ename from sales.bonus) s\n"
+        + "on r.ename=s.ename\n"
+        + "order by r.ename nulls first";
```

Comment:
@hsyuan 
Finally I find a way to "fool" planner to choose the plan I want, basically the plan that pushes sort down to left child of hash join.

The reason that simple join's cheapest plan has a Sort on top of HashJoin, is because the Join's estimated is computed as simple as `left_row_count * right_row_count * 0.15`, 0.15 is selectively. When inputs's row count is small, the selectively plays a bigger role, thus HashJoin gives a small cost, thus Sort is put after Join because Join gives a very small row count estimation.

So I intentionally use UNION to increase input size, which make the  ``left_row_count * right_row_count` plays a bigger role for Join, and then Join produces larger row count estimation. Then finally planner believes push down Sort is cheaper, thus gives the plan I want to see.

Reply:
I know the test query becomes ugly and not easy to check, but I don't know if there is a better idea.

Reply:
It's best actually if current testing suites can allow us mock data, due to cost-based model, its's better to use data rather than wired query to play with it.

I could also check if I can modify the testing framework.

Reply:
How about this:
```sql
select * from
(select distinct a, b from foo) foo
join bar on foo.c = bar.c
order by b desc, a desc; 
```

Reply:
Add an aggregation is not helping. The hash join still uses `left_row_count * right_row_count * selectivity` to estimate cost while the aggreagtion cannot give a big enough `left_row_count`.

Reply:
Problem is solved. I noticed that there are several large tables in `customers` schema. Those large tables can generate plans I want.And if there are future needs, we can always mock new tables for optimization testing.

I pushed a new commit to illustrate which table I am using.

With good input data, I will continue implementation work.